diff --git a/source3/winbindd/idmap_ad.c b/source3/winbindd/idmap_ad.c
index a93c61f54d1..cf2416495cb 100644
--- a/source3/winbindd/idmap_ad.c
+++ b/source3/winbindd/idmap_ad.c
@@ -162,6 +162,13 @@ static TLDAPRC get_attrnames_by_oids(struct tldap_context *ld,
 	}
 
 	TALLOC_FREE(msgs);
+	for (i=0; i<num_oids; i++) {
+		if (names[i] == NULL) {
+			DBG_ERR("Failed to retrieve schema name for "
+				"oid [%s]\n", oids[i]);
+			return TLDAP_FILTER_ERROR;
+		}
+	}
 
 	return TLDAP_SUCCESS;
 }
@@ -228,6 +235,13 @@ static TLDAPRC get_posix_schema_names(struct tldap_context *ld,
 	}
 
 	rc = get_attrnames_by_oids(ld, schema, schema_path, 6, oids, names);
+	if (TLDAP_RC_EQUAL(rc, TLDAP_FILTER_ERROR)) {
+		DBG_ERR("Failed to retrieve schema names for [%s] "
+			"switching to rfc2307.\n", schema_mode);
+		oids = oids_rfc2307;
+		rc = get_attrnames_by_oids(ld, schema, schema_path,
+					   6, oids, names);
+	}
 	TALLOC_FREE(schema_path);
 	if (!TLDAP_RC_IS_SUCCESS(rc)) {
 		TALLOC_FREE(schema);
