From 80460986fc7168bd02701e8240c33f83320766ee Mon Sep 17 00:00:00 2001
From: Andrew Walker <awalker@ixsystems.com>
Date: Mon, 3 May 2021 16:56:05 -0400
Subject: [PATCH] Allow setting SPARSE flag with kernel dosmodes

---
 source3/smbd/dosmode.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/source3/smbd/dosmode.c b/source3/smbd/dosmode.c
index ccfeaca124d..186765e6672 100644
--- a/source3/smbd/dosmode.c
+++ b/source3/smbd/dosmode.c
@@ -1140,10 +1140,6 @@ NTSTATUS file_set_sparse(connection_struct *conn,
 	DEBUG(10,("file_set_sparse: setting sparse bit %u on file %s\n",
 		  sparse, smb_fname_str_dbg(fsp->fsp_name)));
 
-	if (!lp_store_dos_attributes(SNUM(conn))) {
-		return NT_STATUS_INVALID_DEVICE_REQUEST;
-	}
-
 	status = vfs_stat_fsp(fsp);
 	if (!NT_STATUS_IS_OK(status)) {
 		return status;
-- 
2.31.1

