diff --git a/source3/param/service.c b/source3/param/service.c
index 09b8228daad..634ac9e94eb 100644
--- a/source3/param/service.c
+++ b/source3/param/service.c
@@ -129,6 +129,10 @@ int find_service(TALLOC_CTX *ctx, const char *service_in, char **p_service_out)
 
 	iService = lp_servicenumber(*p_service_out);
 
+	if (iService < 0) {
+		iService = load_registry_service(*p_service_out);
+	}
+
 	/* now handle the special case of a home directory */
 	if (iService < 0) {
 		char *phome_dir = get_user_home_dir(ctx, *p_service_out);
